Optimize import/export resolution

hubertp opened the pull request:
### Pull Request Description

This change adds serialization and deserialization of library bindings.
In order to be functional, one needs to first generate IR and
serialize bindings using `--compiled <path-to-library>` command. The bindings
will be stored under the library with `.bindings` suffix.
Bindings are being generated during `buildEngineDistribution` task, thus not 
requiring any extra steps.

When resolving import/exports the compiler will first try to load
module's bindings from cache. If successful, it will not schedule its
imports/exports for immediate compilation, as we always did, but use the
bindings info to infer the dependent modules.

The current change does not make any optimizations when it comes to
compiling the modules, yet. It only delays the actual
compilation/loading IR from cache so that it can be done in bulk.
Further optimizations will come from this opportunity such as parallel
loading of caches or lazily inferring only the necessary modules.

Part of https://github.com/enso-org/enso/issues/5568 work.

### Important Notes

<!--
- Mention important elements of the design.
- Mention any notable changes to APIs.
-->

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed: Enso GUI was tested when built using BOTH
        `./run ide build` and `./run ide watch`.

hubertp committed:
wip: Save and load library bindings

This change adds serialization and deserialization of library bindings.
In order to be functional, one can needs to first generate IR and
serialize bindings `--compiled <path-to-library>` command. The bindings
will be stored under the library with `.bindings` suffix.

When resolving import/exports the compiler will first try to load
module's bindings from cache. If successful, it will not schedule its
imports/exports for immediate compilation, as we always did, but use the
bindings info to infer the dependent modules.

The current change does not make any optimizations when it comes to
compiling the modules, yet. It only delays the actual
compilation/loading IR from cache so that it can be done in bulk.
Further optimizations will come from this opportunity such as parallel
loading of caches or lazilly inferring only the necessary modules.

Other important things to note:
 - deserialization has to take into account dependencies of modules,
   otherwise one may not be able to fully restore abstract module
references
 - bindings map must have a stable serialization uid in order to avoid
   spurious and costly failed cache loading attempts

hubertp committed:
DRY

hubertp committed:
Adapt test to changes

hubertp committed:
Fix invalidation of caches

hubertp committed:
Avoid race-condition

Race-condition could happen (as discovered by the failing test) if we
were serializing sources read from the module rather than the cache
entry. Had to modify the interface for consistency.

JaroslavTulach commented on engine/runtime/src/main/scala/org/enso/compiler/Cache.scala:
```diff
@@ -0,0 +1,463 @@
+  private def saveCacheTo(
+    cacheRoot: TruffleFile,
+    entry: T
+  )(implicit logger: TruffleLogger): Boolean = {
```
Out of curiosity? Is there any reason/benefit of using `implicit logger` over making it normal method argument?

hubertp replied:
runtime - no. It's only that you don't have to pass it around in the code because it is inserted implicitly by the compiler.

JaroslavTulach commented on engine/runtime/src/main/scala/org/enso/compiler/Cache.scala:
```diff
@@ -0,0 +1,463 @@
+  protected def metadataSuffix: String
+
+  protected def messageDigest(): MessageDigest = {
+    new SHA3.Digest224()
```
We both noticed that computing hash occupies significant amount of time during startup. Wouldn't it be simpler and faster to compute MD5? Why do we need anything more complex? We are not concerned about malicious injection of wrong caches, right? Only about accidental corruption of the data. Then anything (even XOR, CRC, etc.) is good enough, I think.

JaroslavTulach replied:
I've just found the [discussion](https://github.com/enso-org/enso/blob/1bc27501e6c8a6018e4c17e68ef4cb4068db15f9/distribution/lib/Standard/Table/0.0.0-dev/src/Data/Table.enso#L118). It is nice to see when such [parallel "inventions"](http://wiki.apidesign.org/wiki/Parallel_Inventions) happen.

JaroslavTulach commented on engine/runtime/src/main/scala/org/enso/compiler/Cache.scala:
```diff
@@ -0,0 +1,463 @@
+package org.enso.compiler
+
+import com.oracle.truffle.api.{TruffleFile, TruffleLogger}
+import org.bouncycastle.jcajce.provider.digest.SHA3
```
Curious: Why do we use bouncycastle and not JDK's own `MessageDigest` implementations?

hubertp replied:
No idea. I just stick to whatever is there and works.

JaroslavTulach commented on engine/runtime/src/main/scala/org/enso/compiler/Cache.scala:
```diff
@@ -0,0 +1,463 @@
+
+}
+
+object Cache {
```
Having a shared access to caches is a great move forward. I would, however, prefer if the new code was written in Java rather than Scala. If this was a new code (as I realized later)...

hubertp replied:
This was a big one but I did it just for you :)

JaroslavTulach commented on engine/runtime/src/main/scala/org/enso/compiler/Compiler.scala:
```diff
@@ -107,28 +108,14 @@ class Compiler(
-        serializationManager.serialize(
+        serializationManager.serializeModule(
           builtins.getModule,
           useGlobalCacheLocations = true // Builtins can't have a local cache.
```
Interesting (but unrelated) comment. Why builtins can't have a local cache? If "local" means inside of the (standard) library which is inside of an Enso distribution (which also contains implementation of all the builtins), then I believe the caches should be there, ready for the 1st usage.

hubertp replied:
Can't remember. The only way to find out is to remove that hardcoded argument :)

JaroslavTulach commented on engine/runtime/src/main/scala/org/enso/compiler/ModuleCache.scala:
```diff
@@ -29,344 +22,44 @@ import scala.util.{Failure, Success, Using}
-    * @return returns the location of the cache if successful, and [[None]] if
-    *         it was unable to save
-    */
-  def save(
```
Huge simplification. Now I see where the content of the newly created `Cache.scala` file comes from.

JaroslavTulach commented on engine/runtime/src/main/scala/org/enso/compiler/SerializationManager.scala:
```diff
@@ -86,7 +90,7 @@ class SerializationManager(compiler: Compiler) {
     *         `true` if `module` has been successfully serialized, `false` otherwise
     */
-  def serialize(
+  def serializeModule(
```
Reasonable rename.

JaroslavTulach commented on engine/runtime/src/main/scala/org/enso/compiler/SerializationManager.scala:
```diff
@@ -129,6 +133,115 @@ class SerializationManager(compiler: Compiler) {
+
+    val task: Callable[Boolean] =
+      doSerializeLibrary(library, useGlobalCacheLocations)
+    if (compiler.context.getEnvironment.isCreateThreadAllowed) {
```
Does it make sense to check `isCreateThreadAllowed` since there is [createSystemThread](https://www.graalvm.org/truffle/javadoc/com/oracle/truffle/api/TruffleLanguage.Env.html#createSystemThread-java.lang.Runnable-java.lang.ThreadGroup-)?

Let's remove the check and always create a *system thread*.

hubertp replied:
Looks like this wasn't updated when the change to system threads was made (GraalVM upgrade).
TBH I would like to change it back so that we don't create system thread. It appears that logs are not being shown so I often do that on my own branch. And the only reason we create system thread is because of native image

JaroslavTulach replied:
I am not sure I understand. We create system threads as some languages don't support multi-threaded access. Also each new `createThread` puts a burden on the interpreter. While system threads are just threads without no change for the generated code. We create system threads as they are "lighter" - not only because of native image.

If the logs aren't shown, then we have to fix it. Is that really related to system threads?

hubertp replied:
> I am not sure I understand. We create system threads as some languages don't support multi-threaded access. Also each new `createThread` puts a burden on the interpreter.

That's not entirely the case for `SerializationManager`. The change from `createThread` to `createSystemThread` was completely dictated by GraalVM upgrade and native-image build failure (potentially being less permissive).

>  While system threads are just threads without no change for the generated code. We create system threads as they are "lighter" - not only because of native image.

OK.

> If the logs aren't shown, then we have to fix it. Is that really related to system threads?

Yes, after the switch we no longer get logs from `TruffleLogger`.

JaroslavTulach replied:
> Yes, after the switch we no longer get logs from TruffleLogger.

Can you formulate that as a high priority issue? E.g. what to run, what log was printed and isn't anymore?

JaroslavTulach commented on engine/runtime/src/main/scala/org/enso/compiler/SerializationManager.scala:
```diff
@@ -146,7 +259,7 @@ class SerializationManager(compiler: Compiler) {
     } else {
-      while (isSerializing(module)) {
+      while (isSerializingModule(module)) {
         Thread.sleep(100)
```
`wait(100)` and a `notifyAll()` when serialization is over would be a nicer that unconditional sleep.

JaroslavTulach commented on engine/runtime/src/main/scala/org/enso/compiler/data/BindingsMap.scala:
```diff
@@ -21,6 +21,10 @@ import scala.annotation.unused
   */
+
+@SerialVersionUID(
+  5568L // stable serialization of bindings
```
I am glad to see _"the pattern"_  spreading.

JaroslavTulach commented on engine/runtime/src/main/scala/org/enso/compiler/phase/ImportResolver.scala:
```diff
@@ -33,67 +33,107 @@ class ImportResolver(compiler: Compiler) {
+    *         a list of all modules which have been inferred from bindings cache and could potentially be compiled lazilly
     */
-  def mapImports(module: Module): List[Module] = {
+  def mapImports(
```
Is this the part that is using the new caches to _"speed something up"_?

hubertp replied:
Not much _speeding up_ yet. Well, that's a lie - there is some. We replace calculations of imports/exports in bindings by loading them from the cache. But effectively all modules that were loaded before are loaded now as well.

JaroslavTulach reviewed (commented):
Shared infrastructure for caching is something other work, including #5068, shall benefit from. Looks good. Few (sometimes unrelated) comments left.

hubertp committed:
Fix unhandled case in equality

Seems like my code managed to trigger a case when comparing an
`AtomConstructor` wrapped in a `Warning` with another
`AtomConstructor`.

Somehow we are already past `equalsWithWarnings` and enter the fallback
specialization.
Not yet sure how to reproduce this in a single test but my weird failing
test in `Integrations_Tests` now succeeds.

hubertp committed:
Steal solution for generating library caches

Originally designed by Dmitry for generating suggestions during build
time, it also works for generating library caches.

hubertp committed:
Ensure modules are ordered properly for codegen

Compiler needs to run ExportResolution's sorting of graph nodes to
ensure the correct order of resolved modules.
Otherwise some exported symbols are resolved before being added to
module's scope and lead to runtime No_Such_Method errors that are hart
to pin down.

Also added an optional way to parse modules in parallel which is turned
off by default. This will allow for easier experimentation on how much
impact that has.

hubertp committed:
We don't have enso.exe on Windows

hubertp committed:
update changelog

hubertp committed:
update docs

hubertp committed:
Windows friendly executable

hubertp committed:
Another attempt to fix Windows build

hubertp committed:
1-to-1 translation to Java

Slowly moving the new code to Java, as per previous discussions.
Added some duplicate API for Java to avoid horrible wrapping with
converters that will quickly get out of hand.

Because of a bug in Frgaal, had to use a simple record-like class rather
than record itself.

hubertp committed:
Drop Java suffix and Scala equivalents

hubertp committed:
Fix test

hubertp committed:
fmt

hubertp committed:
nits

hubertp committed:
More test fixes

JaroslavTulach commented on engine/runtime/src/main/scala/org/enso/compiler/Compiler.scala:
```diff
@@ -241,18 +253,19 @@ class Compiler(
         module
-          .wasLoadedFromCache() && modules.exists(!_.wasLoadedFromCache())
+          .wasLoadedFromCache() && modules
+          .exists(m => !m.wasLoadedFromCache() && !m.isSynthetic)
```
This `!m.isSynthetic` check probably addresses #5608
